### ISSUE TITLE
Update OS used to build .NET package

### ISSUE THREAD

**Issue by bfren** [jedisct1/libsodium#1233]
Currently the build and test jobs for the .NET package run on Ubuntu 16.04 - now Xenial is no longer receiving maintenance updates this should change.

**Comment by bfren** [comment]
The issue is the `run-test-binaries-os-versions` job, which tests compatibility with CentOS 7 and Debian 10.

So far I have used the following for the `build-linux-glibc-xx` jobs:

- Ubuntu 20.04 [fails](https://github.com/bfren/libsodium/actions/runs/3546886904/jobs/5956491551#step:4:12)
- Ubuntu 22.04 [fails](https://github.com/bfren/libsodium/actions/runs/3546806684/jobs/5956380882#step:4:21)
- Debian 10 [fails](https://github.com/bfren/libsodium/actions/runs/3546993537/jobs/5956743031#step:4:12)

Each time the errors are:

```
Run chmod +x .libsodium-builds/linux-x64/Tests
  chmod +x .libsodium-builds/linux-x64/Tests
  .libsodium-builds/linux-x64/Tests
  shell: sh -e {0}
  env:
    DOTNET_SYSTEM_GLOBALIZATION_INVARIANT: 1
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'libsodium' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibsodium: cannot open shared object file: No such file or directory
   at Program.sodium_version_string()
   at Program.Main() in /__w/libsodium/libsodium/.libsodium-test/Program.cs:line 8
/__w/_temp/10cd1428-5f50-4743-9b21-01f1bbd2452a.sh: line 2:    39 Aborted                 (core dumped) .libsodium-builds/linux-x64/Tests
Error: Process completed with exit code 134.
```

**However** using Debian 10 as the build platform (supported until 30 June 2024) means the Debian 10 tests pass (all the other tests pass on all the above) - so the main problem is CentOS 7.

**Comment by jedisct1** [comment]
I'd love to ditch CentOS 7 compatibility, but people are going to complain that it is still supported by the vendor.

**Comment by ektrah** [comment]
.NET unfortunately masks the actual error: It tries to load `libsodium` from a number of places and using different spellings (e.g., by adding a `.so` suffix and/or a `lib` prefix to the library name). If it cannot load and initialize the library, it throws a `DllNotFoundException` saying "liblibsodium: cannot open shared object file: No such file or directory", even if one of the library exists in one of the places and spellings.

The solution to finding the actual error is given in the exception message: "In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable". My guess is that this will show an "Undefined reference to memcpy@GLIBC_2.14" error.

**Comment by ektrah** [comment]
When compiling `libsodium.so` on `ubuntu:22.04`, loading that file fails on both `centos:7` and `debian:10` with the following error:

```
checking for version `GLIBC_2.3' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.33' in file /lib64/libc.so.6 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
/lib64/libc.so.6: error: version lookup error: version `GLIBC_2.33' not found (required by /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so) (fatal)
```

This is independent of whether `__asm__(".symver memcpy,memcpy@GLIBC_2.2.5");` is present or not.

When compiling on `ubuntu:20.04` or `ubuntu:18.04` and loading on `centos:7`:

```
checking for version `GLIBC_2.3' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.2.5' in file /lib64/libpthread.so.0 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.14' in file /lib64/libc.so.6 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.3.4' in file /lib64/libc.so.6 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.4' in file /lib64/libc.so.6 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
checking for version `GLIBC_2.25' in file /lib64/libc.so.6 [0] required by file /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so [0]
/lib64/libc.so.6: error: version lookup error: version `GLIBC_2.25' not found (required by /__w/libsodium/libsodium/.libsodium-builds/linux-x64/libsodium.so) (fatal)
```

When compiling on `ubuntu:16.04`, everything works (without `__asm__(".symver memcpy,memcpy@GLIBC_2.2.5");`).

So it seems that, until CentOS 7 reaches end-of-life (or .NET drops support for it), we're stuck with compiling `libsodium.so` for .NET on `ubuntu:16.04`...

---

glibc versions of distributions:

| Distribution | glibc version | EOL |
|-|-|-|
| CentOS 7.7.1908 | 2.17 | 2024-06 |
| Debian 10 | 2.28 | 2022-09 |
| Ubuntu 16.04 | 2.23 | 2026-04 |
| Ubuntu 18.04 | 2.27 | 2028-04 |
| Ubuntu 20.04 | 2.31 | 2030-04 |
| Fedora 35 | 2.34 | 2022-12 |
| Fedora 36 | 2.53 | 2023-05 |

---

[libc compatibility](https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md) of .NET itself:

> * x64: [glibc](https://www.gnu.org/software/libc/) 2.17 (from CentOS 7)
> * Arm32, Arm64: [glibc](https://www.gnu.org/software/libc/) 2.23 (from Ubuntu 16.04)
> * Alpine (x64 and Arm64): [musl](https://musl.libc.org/) 1.2.2 (from Alpine 3.13)

**Comment by jedisct1** [comment]
Another option is to compile with Zig, that can cross-compile to many targets, including specific libc versions:

Example:

```sh
zig build -Drelease-fast -Dtarget=x86_64-linux-gnu.2.17
```

**Comment by jedisct1** [comment]
Looks like this is the best way to go.

`dotnet-core.yml` is now using `zig` to build Linux binaries.

Compilation can thus be done on current Linux distributions, while targeting older glibc versions. So, we target glibc 2.17 for x86_64 and glibc 2.23 for ARM, according to your table.

Compilation to aarch64 and arm with musl was fixed by the way (previously, the aarch64 build was actually an x86_64 one).

**Comment by jedisct1** [comment]
Adding Windows ARM would probably also be nice.

`zig build -Drelease-fast -Dtarget=aarch64-windows` gives us a `DLL` and a `.LIB` Windows ARM.  But I guess this is impossible to test on GitHub CI, right?

**Comment by bfren** [comment]
Is 1.0.18.3 waiting on this?

**Comment by jedisct1** [comment]
1.0.18.3 has been published :)

**Comment by ektrah** [comment]
Some unit tests in NSec that passed with 1.0.18.2 [are now failing](https://github.com/ektrah/nsec/actions/runs/3669008323 ) with 1.0.18.3 ☹️

It looks like something changed in or around `crypto_aead_aes256gcm_decrypt`?

**Comment by ektrah** [comment]
And it seems the pre-compiled binary [built for `linux-musl-arm`](https://github.com/jedisct1/libsodium/blob/408125a72b5cbf0ccd9e478dae6b90f8737d3ee7/.github/workflows/dotnet-core.yml#L237-L240) isn't actually [included in the NuGet package](https://github.com/jedisct1/libsodium/blob/408125a72b5cbf0ccd9e478dae6b90f8737d3ee7/packaging/dotnet-core/libsodium.pkgproj#L28-L36)...

**Comment by jedisct1** [comment]
Ah, looks like that has to be added to `libsodium.pkgproj`.

**Comment by bfren** [comment]
Ahh, that must also have got clobbered like the version bump I did.

**Comment by bfren** [comment]
> Some unit tests in NSec that passed with 1.0.18.2 [are now failing](https://github.com/ektrah/nsec/actions/runs/3669008323) with 1.0.18.3 ☹️
> 
> It looks like something changed in or around `crypto_aead_aes256gcm_decrypt`?

Is this to do with 408125a72b5cbf0ccd9e478dae6b90f8737d3ee7?

**Comment by jedisct1** [comment]
> > Some unit tests in NSec that passed with 1.0.18.2 [are now failing](https://github.com/ektrah/nsec/actions/runs/3669008323) with 1.0.18.3 ☹️
> > It looks like something changed in or around `crypto_aead_aes256gcm_decrypt`?
> 
> Is this to do with [408125a](https://github.com/jedisct1/libsodium/commit/408125a72b5cbf0ccd9e478dae6b90f8737d3ee7)?

Very unlikely. This code doesn't exist in `stable`.

**Comment by jedisct1** [comment]
On a failed decryption, the message buffer is now filled with `0xd0` instead of `0x00` (in the tradition of OpenBSD, making it more obvious that something went wrong instead of an all-null content).

The NSec test should check that the output buffer doesn't match the plaintext rather than expect a specific output on failure.

**Comment by jedisct1** [comment]
What do the other failing tests do?

**Comment by ektrah** [comment]
The simplest possible test case fails if `clen >= 256`.

**Comment by jedisct1** [comment]
Oops! That should be fixed by a5ea347381991c7c4c0ca9701428e53677c65f8a 

Our test suite was missing tests with long inputs. This case is now tested.
